File: dotnet/Lexy/Lexy.Compiler/Language/NodesWalker.cs
using System;
using System.Collections.Generic;

namespace Lexy.Compiler.Language
{
    internal static class NodesWalker
    {
        public static IEnumerable<T> WalkWithResult<T>(IEnumerable<INode> nodes, Func<INode, T> action)
        {
            if (nodes == null) throw new ArgumentNullException(nameof(nodes));
            if (action == null) throw new ArgumentNullException(nameof(action));

            var result = new List<T>();
            WalkWithResult(nodes, action, result);

            return result;
        }

        private static void WalkWithResult<T>(INode node, Func<INode, T> action, IList<T> result)
        {
            var actionResult = action(node);
            if (actionResult != null)
            {
                result.Add(actionResult);
            }

            var children = node.GetChildren();

            WalkWithResult(children, action, result);
        }

        private static void WalkWithResult<T>(IEnumerable<INode> nodes, Func<INode, T> action, IList<T> result)
        {
            foreach (var node in nodes)
            {
                WalkWithResult(node, action, result);
            }
        }
    }
}

File: typescipt/src/language/nodesWalker.ts


internal static class NodesWalker {
   public static walk(nodes: Array<INode>, action: Action<INode>): void {
     if (nodes == null) throw new Error(nameof(nodes));
     if (action == null) throw new Error(nameof(action));

     foreach (let node in nodes) Walk(node, action);
   }

   public static walk(node: INode, action: Action<INode>): void {
     if (node == null) throw new Error(nameof(node));
     if (action == null) throw new Error(nameof(action));

     action(node);

     let children = node.GetChildren();
     Walk(children, action);
   }

   public static walk(nodes: Array<INode>, function: Func<INode, bool>): boolean {
     if (nodes == null) throw new Error(nameof(nodes));
     if (function == null) throw new Error(nameof(function));

     foreach (let node in nodes) {
       if (!function(node)) return false;

       let children = node.GetChildren();
       if (!Walk(children, function)) return false;
     }

     return true;
   }

   public static walkWithResult<T>(nodes: Array<INode>, action: Func<INode, T>): Array<T> {
     if (nodes == null) throw new Error(nameof(nodes));
     if (action == null) throw new Error(nameof(action));

     let result = new Array<T>();
     WalkWithResult(nodes, action, result);

     return result;
   }

   private static walkWithResult<T>(node: INode, action: Func<INode, T>, result: Array<T>): void {
     let actionResult = action(node);
     if (actionResult != null) result.Add(actionResult);

     let children = node.GetChildren();

     WalkWithResult(children, action, result);
   }

   private static walkWithResult<T>(nodes: Array<INode>, action: Func<INode, T>, result: Array<T>): void {
     foreach (let node in nodes) WalkWithResult(node, action, result);
   }
}
